File: easy_entry/easy_entry/report/accounts_receivable_items/accounts_receivable_items.py
import frappe
from frappe import _
from frappe.utils import flt, today


def execute(filters=None):
	filters = frappe._dict(filters or {})
	if not filters.as_of_date:
		filters.as_of_date = today()

	columns = get_columns()
	data = get_data(filters)
	summary = get_summary(data)

	return columns, data, None, None, summary


def get_columns():
	return [
		{
			"fieldname": "customer",
			"label": _("العميل / Customer"),
			"fieldtype": "Link",
			"options": "Customer",
			"width": 150,
		},
		{
			"fieldname": "customer_name",
			"label": _("الاسم / Name"),
			"fieldtype": "Data",
			"width": 155,
		},
		{
			"fieldname": "outstanding_amount",
			"label": _("المستحق / Outstanding"),
			"fieldtype": "Currency",
			"width": 135,
		},
		{
			"fieldname": "invoice",
			"label": _("الفاتورة / Invoice"),
			"fieldtype": "Link",
			"options": "Sales Invoice",
			"width": 135,
		},
		{
			"fieldname": "posting_date",
			"label": _("التاريخ / Date"),
			"fieldtype": "Date",
			"width": 90,
		},
		{
			"fieldname": "item_code",
			"label": _("الصنف / Item"),
			"fieldtype": "Link",
			"options": "Item",
			"width": 105,
		},
		{
			"fieldname": "item_name",
			"label": _("اسم الصنف / Item Name"),
			"fieldtype": "Data",
			"width": 160,
		},
		{
			"fieldname": "qty",
			"label": _("الكمية / Qty"),
			"fieldtype": "Float",
			"width": 60,
		},
		{
			"fieldname": "rate",
			"label": _("السعر / Rate"),
			"fieldtype": "Currency",
			"width": 95,
		},
		{
			"fieldname": "amount",
			"label": _("الإجمالي / Amount"),
			"fieldtype": "Currency",
			"width": 105,
		},
	]


def get_data(filters):
	conditions = [
		"si.docstatus = 1",
		"si.outstanding_amount > 0.009",
		"si.posting_date <= %(as_of_date)s",
	]
	params = {"as_of_date": filters.as_of_date}

	if filters.get("company"):
		conditions.append("si.company = %(company)s")
		params["company"] = filters.company

	if filters.get("customer"):
		conditions.append("si.customer = %(customer)s")
		params["customer"] = filters.customer

	where = " AND ".join(conditions)

	rows = frappe.db.sql(
		f"""
		SELECT
			si.customer,
			si.customer_name,
			si.name AS invoice,
			si.posting_date,
			si.grand_total,
			si.outstanding_amount,
			sii.item_code,
			sii.item_name,
			sii.qty,
			sii.rate,
			sii.amount
		FROM `tabSales Invoice` si
		JOIN `tabSales Invoice Item` sii ON sii.parent = si.name
		WHERE {where}
		ORDER BY si.customer, si.posting_date, si.name, sii.idx
		""",
		params,
		as_dict=True,
	)

	if not rows:
		return []

	# Aggregate customer totals first (one pass)
	customer_totals = {}
	customer_names = {}
	for row in rows:
		key = row.customer
		grand_total = flt(row.grand_total)
		item_outstanding = (
			flt(row.outstanding_amount) * flt(row.amount) / grand_total
			if grand_total
			else flt(row.amount)
		)
		customer_totals[key] = customer_totals.get(key, 0.0) + item_outstanding
		customer_names[key] = row.customer_name or row.customer

	# Build output: group header per customer (with total), then detail rows
	data = []
	current_customer = None

	for row in rows:
		if row.customer != current_customer:
			current_customer = row.customer

			# Customer header row: shows ID, name, and total outstanding
			data.append(
				{
					"customer": row.customer,
					"customer_name": customer_names[row.customer],
					"invoice": None,
					"posting_date": None,
					"item_code": None,
					"item_name": None,
					"qty": None,
					"rate": None,
					"amount": None,
					"outstanding_amount": customer_totals[row.customer],
					"is_group": 1,
					"indent": 0,
				}
			)

		grand_total = flt(row.grand_total)
		item_outstanding = (
			flt(row.outstanding_amount) * flt(row.amount) / grand_total
			if grand_total
			else flt(row.amount)
		)

		data.append(
			{
				"customer": None,
				"customer_name": None,
				"invoice": row.invoice,
				"posting_date": row.posting_date,
				"item_code": row.item_code,
				"item_name": row.item_name,
				"qty": flt(row.qty),
				"rate": flt(row.rate),
				"amount": flt(row.amount),
				"outstanding_amount": item_outstanding,
				"is_group": 0,
				"indent": 1,
			}
		)

	return data


def get_summary(data):
	header_rows = [r for r in data if r.get("is_group")]
	total = sum(flt(r.get("outstanding_amount") or 0) for r in header_rows)
	customer_count = len(header_rows)

	return [
		{
			"label": _("عدد العملاء / Customers"),
			"value": customer_count,
			"datatype": "Int",
			"color": "blue",
		},
		{
			"label": _("إجمالي المستحق / Total Outstanding"),
			"value": total,
			"datatype": "Currency",
			"color": "orange",
		},
	]

File: easy_entry/easy_entry/report/accounts_receivable_items/accounts_receivable_items.js
frappe.query_reports["Accounts Receivable Items"] = {
	filters: [
		{
			fieldname: "company",
			label: __("الشركة / Company"),
			fieldtype: "Link",
			options: "Company",
			default: frappe.defaults.get_default("company"),
		},
		{
			fieldname: "customer",
			label: __("العميل / Customer"),
			fieldtype: "Link",
			options: "Customer",
		},
		{
			fieldname: "as_of_date",
			label: __("حتى تاريخ / As of Date"),
			fieldtype: "Date",
			default: frappe.datetime.get_today(),
			reqd: 1,
		},
	],

	onload: function (report) {
		// Inject mobile-friendly styles
		if (!document.getElementById("ar-report-styles")) {
			const style = document.createElement("style");
			style.id = "ar-report-styles";
			style.textContent = `
				/* Make the report area scrollable on mobile */
				.page-content-wrapper .datatable {
					overflow-x: auto !important;
					-webkit-overflow-scrolling: touch;
				}

				/* Compact cells on mobile */
				@media (max-width: 768px) {
					.dt-cell__content {
						font-size: 12px !important;
						padding: 4px 6px !important;
					}
					.dt-cell__resize-handle { display: none !important; }
					.report-summary {
						flex-direction: column !important;
						gap: 8px !important;
					}
					.report-summary-item {
						padding: 10px 14px !important;
						border-radius: 8px !important;
					}
					.report-summary-value {
						font-size: 20px !important;
					}
					/* Customer header rows — bigger touch targets on mobile */
					.dt-row[data-is-group="1"] .dt-cell__content,
					.dt-row .dt-cell__content[data-is-group="1"] {
						padding: 6px 8px !important;
					}
				}

				/* Summary cards — always look good */
				.report-summary {
					display: flex;
					gap: 12px;
					flex-wrap: wrap;
					margin-bottom: 16px;
				}
				.report-summary-item {
					flex: 1 1 180px;
				}
				.report-summary-value.text-blue {
					color: #2563eb !important;
				}
				.report-summary-value.text-orange {
					color: #b45309 !important;
				}

				/* Tighten the filter bar on mobile */
				@media (max-width: 576px) {
					.filter-field { width: 100% !important; margin-bottom: 6px; }
					.page-actions { flex-wrap: wrap; gap: 6px; }
				}
			`;
			document.head.appendChild(style);
		}
	},

	formatter: function (value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);

		if (!data) return value;

		if (data.is_group) {
			let inner = "";

			if (column.fieldname === "customer") {
				inner = value
					? `<span style="font-weight:700;color:#92400e;font-size:0.88em;letter-spacing:0.04em;">${value}</span>`
					: "";
			} else if (column.fieldname === "customer_name") {
				inner = value
					? `<span style="font-weight:700;color:#1c1917;font-size:0.97em;">${value}</span>`
					: "";
			} else if (column.fieldname === "outstanding_amount") {
				inner = value
					? `<span style="display:inline-block;background:#fef3c7;color:#92400e;border:1.5px solid #f59e0b;border-radius:5px;padding:1px 10px;font-weight:800;font-size:0.93em;white-space:nowrap;">${value}</span>`
					: "";
			} else {
				inner = value || "";
			}

			return `<div style="background:#fef3c7;margin:-3px -8px;padding:3px 8px;min-height:22px;border-top:1.5px solid #fde68a;border-bottom:1.5px solid #fde68a;">${inner}</div>`;
		} else {
			if (column.fieldname === "outstanding_amount" && value) {
				return `<span style="color:#b45309;font-weight:600;">${value}</span>`;
			}
			if (column.fieldname === "invoice" && value) {
				return `<span style="color:#2563eb;">${value}</span>`;
			}
			return value;
		}
	},
};
